doc/man3/EVP_MAC.pod

```diff
@@ -0,0 +1,235 @@
+=pod
+
+=head1 NAME
+
+EVP_MAC, EVP_MAC_CTX, EVP_MAC_CTX_new, EVP_MAC_CTX_new_id, EVP_MAC_CTX_free,
+EVP_MAC_CTX_copy, EVP_MAC_CTX_mac, EVP_MAC_reset, EVP_MAC_size, EVP_MAC_init,
+EVP_MAC_update, EVP_MAC_final, EVP_MAC_ctrl, EVP_MAC_ctrl_str, EVP_MAC_str2ctrl,
+EVP_MAC_hex2ctrl, EVP_MAC_nid, EVP_MAC_name, EVP_add_mac, EVP_get_macbyname,
+EVP_get_macbynid, EVP_get_macbyobj - EVP MAC routines
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ typedef struct evp_mac_st EVP_MAC;
+ typedef struct evp_mac_ctx_st EVP_MAC_CTX;
+
+ EVP_MAC_CTX *EVP_MAC_CTX_new(const EVP_MAC *mac);
+ EVP_MAC_CTX *EVP_MAC_CTX_new_id(int nid);
+ void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx);
+ int EVP_MAC_CTX_copy(EVP_MAC_CTX *dest, EVP_MAC_CTX *src);
+ const EVP_MAC *EVP_MAC_CTX_mac(EVP_MAC_CTX *ctx);
+ int EVP_MAC_reset(EVP_MAC_CTX *ctx);
+ size_t EVP_MAC_size(EVP_MAC_CTX *ctx);
+ int EVP_MAC_init(EVP_MAC_CTX *ctx);
+ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen);
+ int EVP_MAC_final(EVP_MAC_CTX *ctx, unsigned char *out, size_t *poutlen);
+ int EVP_MAC_ctrl(EVP_MAC_CTX *ctx, int cmd, ...);
+ int EVP_MAC_ctrl_str(EVP_MAC_CTX *ctx, const char *type, const char *value);
+ int EVP_MAC_str2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_hex2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_nid(const EVP_MAC *mac);
+ const char *EVP_MAC_name(const EVP_MAC *mac);
+ int EVP_add_mac(const EVP_MAC *mac);
+ const EVP_MAC *EVP_get_macbyname(const char *name);
+ const EVP_MAC *EVP_get_macbynid(int nid);
+ const EVP_MAC *EVP_get_macbyobj(const ASN1_OBJECT *o);
+
+=head1 DESCRIPTION
+
+These types and functions help the application to calculate MACs of
+different types and with different underlying algorithms if there are
+any.
+
+=head2 Types
+
+B<EVP_MAC> is a type that holds the implementation of a MAC.
+
+B<EVP_MAC_CTX> is a context type that holds internal information used
+to compute MACs from the input.
+
+=head2 Context manipulation functions
+
+EVP_MAC_CTX_new() creates a new context for the MAC type C<mac>.
+EVP_MAC_CTX_new_id() creates a new context for the numerical MAC
+identity <nid>.  The created context can then be used with most other
+functions described here.
+
+EVP_MAC_CTX_free() frees the contents of the context as well as the
+context itself.  B<NULL> is a valid parameter, for which this function
+is a no-op.
+
+EVP_MAC_CTX_copy() makes a deep copy of the C<src> context to the
+C<dest> context.  The C<dest> context I<must> have been created before
+calling this function.
+
+EVP_MAC_CTX_mac() returns the B<EVP_MAC> associated with the context
+C<ctx>.
+
+=head2 Computing functions
+
+EVP_MAC_init() sets up the context with information given through
+diverse controls.  Implementations may vary on this, some of them may
+do their initialization in the controls themselves.  However, this
+should be called at the very least before calling EVP_MAC_update() and
+EVP_MAC_final().
```

Comment:
This reads as, you should call it, but we're not telling you when, only that it needs to be before update, it might need to be before or after ctrl(), and you need to figure that out on your own. I wonder why it needs to be called at all, why isn't EVP_MAC_CTX_new() doing it already?

I don't think we should support reusing, so I don't see a need for this init function.

Reply:
EVP_MAC_CTX_new only creates and initializes the MAC level context, but not the underlying computation algo if there is one. EVP_MAC_init does the latter, after that algo has been specified.

So the call sequence would typically be:

EVP_MAC_CTX_new
EVP_MAC_ctrl (enough times)
EVP_MAC_init
EVP_MAC_update (one or more)
EVP_MAC_final

Reply:
Note that EVP_PKEY_derive_init() does it just the other way around, init first. And as #7497 points out, the init function might not do what you expect to reuse it.

Reply:
EVP_PKEY is a mixture of things, so it's quite hard to draw conclusions from one quick example.

In the MAC case, the driving question is how to initialize an underlying computation algorithm without knowing what that algorithm will be.

Reply:
Thinking a bit more about it, there are definitely controls that can be called after init.  It all depends on what they affect.

Reply:
I take that back...  all controls defined so far affect things that must be in place before the MAC is initialized.

Reply:
What do you think about internally calling this init function in update or final if it wasn't called yet?